base_directory_files_download/models/ir_filesystem_file.py

```diff
@@ -0,0 +1,57 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Onestein (<http://www.onestein.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+import os
+
+from odoo import api, fields, models, _
+from odoo.exceptions import Warning
+from odoo.tools import human_size
+
+_logger = logging.getLogger(__name__)
+
+
+class IrFilesystemDirectoryLine(models.TransientModel):
+    _name = 'ir.filesystem.file'
+
+    name = fields.Char(required=True)
+    filename = fields.Char()
+    file_content = fields.Binary(compute='_compute_file')
+    stored_filename = fields.Char()
+    directory_id = fields.Many2one(
+        'ir.filesystem.directory',
+        string='Directory'
+    )
+
+    @api.multi
+    def _file_read(self, fname, bin_size=False):
+
+        def file_not_found(fname):
+            raise Warning(_(
+                '''Error while reading file %s.
+                Maybe it was removed or permission is changed.
+                Please refresh the list.''' % fname))
+
+        self.ensure_one()
+        r = ''
+        directory = self.directory_id.get_dir() or ''
+        full_path = directory + fname
+        if not (directory and os.path.isfile(full_path)):
```

Comment:
what a shame I missed this in the first review: There's no code that prevents me from creating a record of this type with `stored_filename` equal to `../../../etc/passwd` or the like. Use `normpath` and maybe also `realpath` and check that the result is contained by the directory the file record is linked to